Userland/Libraries/LibCompress/LZW.h

```diff
@@ -17,23 +17,85 @@
 
 namespace Compress::LZW {
 
-template<InputBitStream InputStream>
-class Decoder {
-private:
-    static constexpr int max_code_size = 12;
+namespace Details {
 
+class Base {
 public:
-    explicit Decoder(MaybeOwned<InputStream> lzw_stream, u8 min_code_size, i32 offset_for_size_change = 0)
-        : m_bit_stream(move(lzw_stream))
-        , m_code_size(min_code_size)
+    u16 add_control_code()
+    {
+        u16 const control_code = m_code_table.size();
+        m_code_table.append(Vector<u8> {});
+        m_original_code_table.append(Vector<u8> {});
+        if (m_code_table.size() >= m_table_capacity && m_code_size < max_code_size) {
+            ++m_code_size;
+            ++m_original_code_size;
+            m_table_capacity *= 2;
+        }
+        return control_code;
+    }
+
+    void reset()
+    {
+        m_code_table.clear();
+        m_code_table.extend(m_original_code_table);
+        m_code_size = m_original_code_size;
+        m_table_capacity = AK::exp2<u32>(m_code_size);
+    }
+
+protected:
+    static constexpr int max_code_size = 12;
+
+    Base(u8 min_code_size, i32 offset_for_size_change)
+        : m_code_size(min_code_size)
         , m_original_code_size(min_code_size)
         , m_table_capacity(AK::exp2<u32>(min_code_size))
         , m_offset_for_size_change(offset_for_size_change)
-
     {
         init_code_table();
     }
 
+    void init_code_table()
+    {
+        m_code_table.ensure_capacity(m_table_capacity);
+        for (u16 i = 0; i < m_table_capacity; ++i) {
+            m_code_table.unchecked_append({ (u8)i });
+        }
+        m_original_code_table = m_code_table;
+    }
+
+    void extend_code_table(Vector<u8> const& entry)
+    {
+        if (entry.size() > 1 && m_code_table.size() < 4096) {
+            m_code_table.append(entry);
+            if (m_code_table.size() >= (m_table_capacity + m_offset_for_size_change) && m_code_size < max_code_size) {
+                ++m_code_size;
+                m_table_capacity *= 2;
+            }
+        }
+    }
+
+    Vector<Vector<u8>> m_code_table {};
+    Vector<Vector<u8>> m_original_code_table {};
+
+    u8 m_code_size { 0 };
+    u8 m_original_code_size { 0 };
+
+    u32 m_table_capacity { 0 };
+    i32 m_offset_for_size_change {};
+};
+
+}
+
+template<InputBitStream InputStream>
+class Decoder : Details::Base {
```

Comment:
If we decide to make this a `Stream` in the future, inheriting here (as opposed to, say, composition) might pose problems. Or at least I think so, because for whatever reason I haven't inherited from `LzmaState` in `LzmaCompressor` and `LzmaDecompressor`.

Reply:
I would rather not make a change based on the *possibility* of an issue for something we *might* do in the future. I don't even understand why that would be an issue.

Reply:
> I don't even understand why that would be an issue.

I don't understand why either, but C++ often works in mysterious ways. I just remember that I haven't inherited from the common state for one reason or another.

In any case, definitely not a blocker, just something to keep in mind.

Reply:
Ok, let's merge figure out that later then. We might even want to split `LzwState`  in the future, the data structure is really not adapted for compression.